### PR TITLE
3 types of user authentication

### PR DESCRIPTION
**Pull request by junaidgirkar**
3 user models created: Student, Teacher and Librarian. Registration process available for all 3 and a login page which returns token on user credentials.

### CODE REVIEW

**Comment by SakshiUppoor** [comment]
Please maintain everything in libraryApp only so that file structure remains the same as mentioned in the readme

File: libraryBackend/libraryApp/api.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+        serializer = RegisterSerializer1(users, many=True)
+        return Response(serializer.data)
+
+    def post(self, request, *args, **kwargs):
+        serializer = self.get_serializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        user = serializer.save()
+        return Response(
+            {
+                "user": UserSerializer(
```
Comment (Harshvartak) [inline libraryBackend/libraryApp/api.py]: UserSerializer not defined error thrown for user signup, the user gets registered but no response

File: libraryBackend/libraryApp/models.py [hunk 2]
```diff
@@ -1,3 +1,97 @@
...
+
+        return self.email
+
+
+class Student(User):
+    user = models.OneToOneField(User, on_delete=models.CASCADE, parent_link=True)
+    user.is_student = True
+    user.is_teacher = False
+    user.is_librarian = False
+    department = models.CharField(max_length=40)
```
Comment (Harshvartak) [inline libraryBackend/libraryApp/models.py]: 1. Department Choices to be present
2. SAP validators in the backend
3. SAP as a required field for the Student and Teacher model, Librarian ID for Librarian model, also change blank to False for the mentioned fields

File: libraryBackend/libraryApp/serializers.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+class StudentRegisterSerializer1(serializers.ModelSerializer):
+    class Meta:
+        model = Student
+        fields = (
+            "id",
+            "email",
+            "first_name",
+            "last_name",
+            "sap_id",
+            "department",
```
Comment (Harshvartak) [inline libraryBackend/libraryApp/serializers.py:33]: Graduation Date Missing

File: libraryBackend/libraryApp/api.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    if username is None or password is None:
+        return Response(
+            {"error": "Please provide both username and password"},
+            status=HTTP_400_BAD_REQUEST,
+        )
+    user = authenticate(username=username, password=password)
+    if not user:
+        return Response({"error": "Invalid Credentials"}, status=HTTP_404_NOT_FOUND)
+    token, _ = Token.objects.get_or_create(user=user)
+    return Response({"token": token.key}, status=HTTP_200_OK)
```
Comment (Harshvartak) [inline libraryBackend/libraryApp/api.py]: Make the Login Function Swagger Doc Compatible

**Review by Harshvartak — CHANGES_REQUESTED** [review]

**Review by Harshvartak — APPROVED** [review]

**Pull request merged**